src/VerletList.cpp

```diff
@@ -84,25 +86,180 @@ namespace espressopp {
   
   void VerletList::rebuild()
   {
+    timer.reset();
+    real currTime = timer.getElapsedTime();
+
     //real cutVerlet = cut + getSystem() -> getSkin();
     cutVerlet = cut + getSystem() -> getSkin();
     cutsq = cutVerlet * cutVerlet;
     
     vlPairs.clear();
 
-    // add particles to adress zone
-    CellList cl = getSystem()->storage->getRealCells();
-    LOG4ESPP_DEBUG(theLogger, "local cell list size = " << cl.size());
-    for (CellListAllPairsIterator it(cl); it.isValid(); ++it) {
-      checkPair(*it->first, *it->second);
-      LOG4ESPP_DEBUG(theLogger, "checking particles " << it->first->id() << " and " << it->second->id());
+    if(useBuffers) {
+      rebuildUsingBuffers(exList.size(), useSOA);
     }
-    
+    else
+    {
+      // add particles to adress zone
+      CellList cl = getSystem()->storage->getRealCells();
+      LOG4ESPP_DEBUG(theLogger, "local cell list size = " << cl.size());
+      for (CellListAllPairsIterator it(cl); it.isValid(); ++it) {
+        checkPair(*it->first, *it->second);
+        LOG4ESPP_DEBUG(theLogger, "checking particles " << it->first->id() << " and " << it->second->id());
+      }
+    }
+
     builds++;
+    timeRebuild += timer.getElapsedTime() - currTime;
     LOG4ESPP_DEBUG(theLogger, "rebuilt VerletList (count=" << builds << "), cutsq = " << cutsq
                  << " local size = " << vlPairs.size());
   }
-  
+
+  /*-------------------------------------------------------------*/
+
+  template< bool USE_EXCLUSION_LIST, bool USE_SOA >
+  void VerletList::_rebuildUsingBuffers()
+  {
+    const CellList& realCells = getSystem()->storage->getRealCells();
+    const size_t numRealCells = realCells.size();
+
+    // stores the range of neighbor particles belonging to cell i: c_range[i] to c_range[i+1]
+    std::vector<int> c_range;
+    c_range.reserve(numRealCells+1);
+
+    // get the number of particles in all neighbor cells
+    size_t c_reserve = 0;
+    c_range.push_back(c_reserve);
```

Comment:
Why do you have to push `0` as the first element?

Reply:
This initially stored the range of neighbor particles belonging to icell, `size_t start=c_range[icell], end=c_range[icell+1];`. I've now changed it to `size_t end=c_range[icell];` and removed the need for 0 as first element of `c_range`.